src/julia_gc.cc

```diff
@@ -210,204 +126,471 @@ static inline void * align_ptr(void * p)
+        _data = alloc<T>(cap);
+    }
+    static ArrayList<T> * make(Int cap)
+    {
+        ArrayList<T> * result = (ArrayList<T> *)alloc<ArrayList<T> >();
+        result->init(cap);
+        return result;
+    }
+    static void destroy(ArrayList<T> * array)
+    {
+        array->deinit();
+        dealloc(array);
+    }
+    void deinit()
+    {
+        dealloc(_data);
+        _len = _cap = 0;
+        _data = NULL;
+    }
+    void add(T item)
+    {
+        if (_len == _cap) {
+            Int newcap = _cap ? _cap * 2 : 1;
+            T * data = alloc<T>(newcap);
+            if (_data)
+                memcpy(data, _data, newcap * sizeof(T));
+            dealloc(_data);
+            _data = data;
+            _cap = newcap;
         }
-        else if (c > 0) {
-            treap_insert(&R(t), val);
-            if (R(t)->prio > t->prio) {
-                treap_rot_left(treap);
+        _data[_len++] = item;
+    }
+    template <typename Cmp>
+    void sort()
+    {
+        T * in = alloc<T>(_len);
+        T * out = alloc<T>(_len);
+        if (_data)
+            memcpy(in, _data, sizeof(T) * _len);
+        Int step = 1;
+        while (step < _len) {
+            for (Int i = 0; i < _len; i += step * 2) {
+                Int p = i, l = i, r = i + step, lmax = l + step,
+                    rmax = r + step;
+                if (rmax > _len)
+                    rmax = _len;
+                if (lmax > _len)
+                    lmax = _len;
+                while (l < lmax && r < rmax) {
+                    int c = Cmp::cmp(in[l], in[r]);
+                    if (c < 0) {
+                        out[p++] = in[l++];
+                    }
+                    else {
+                        out[p++] = in[r++];
+                    }
+                }
+                while (l < lmax) {
+                    out[p++] = in[l++];
+                }
+                while (r < rmax) {
+                    out[p++] = in[r++];
+                }
             }
+            T * tmp = in;
+            in = out;
+            out = tmp;
+            step += step;
         }
+        dealloc(_data);
+        dealloc(out);
+        _data = in;
+        _cap = _len;    // we allocated only _len items for 'in'.
     }
-}
-
-static void treap_delete_node(treap_t ** treap)
-{
-    for (;;) {
-        treap_t * t = *treap;
-        if (L(t) == NULL) {
-            *treap = R(t);
-            free_treap(t);
-            break;
+};
+
+// conservative estimate: 2 * log(max(UInt)+1) for alpha ~= 1.0
+static const UInt MaxTreeDepth = 2 * (sizeof(UInt) * 8);
+static int        height_to_size_init = 0;
+// For scapegoat trees with balance factor alpha:
+// height_to_size: d -> (1/alpha) ^ d
+static Int height_to_size[MaxTreeDepth];
+
+template <typename T, typename Cmp = T>
+class BalancedTree : public StdAlloc {
+    // A scapegoat tree. It is an easy to implement balanced
+    // tree, with amortized O(log n) insertion and deletion
```

Comment:
Could you briefly explain why we need to replace the treap code with yet another new data structure?

Reply:
The treap code technically relies on a random number generation; while it is very unlikely to fail, I prefer deterministic behavior. Scapegoat trees are the easiest to implement deterministic balanced tree. I could have used `std::set` and `std::map`, but they again rely on the missing C++ runtime, so I plugged in some code that I had already written for something else.

Reply:
OK, that explains why the shift away from a treap. I also agree that we shouldn't use `std::set` and `std::map`.